### ISSUE TITLE
Clang analyzer can't detect null dereferencing

### ISSUE THREAD

**Issue by Karina5005** [llvm/llvm-project#56793]
I came across a simple example of dereferencing a null pointer, but clang tidy doesn't detect this problem. 
```
class ClassA {
private:
    int *a = new int(5);

public:
    int *get() {
        return a;
    }
    void set(int* newA) {
        a = newA;
    }
};

int main() {
    ClassA classA = ClassA();
    classA.set(nullptr);
    return *classA.get();
}
```
The problem is in the getter, if you make the field public and get it directly, everything works correctly. 
Why is that? Is there any restriction on the detection of such situations or is it a bug? Or maybe I'm doing something wrong?

**Comment by dwblaikie** [comment]
Does `clang-tidy` run the Clang Static Analyzer? I thought it mostly/entirely just ran normal clang diagnostics and specific `clang-tidy` checks?

**Comment by EugeneZelenko** [comment]
@dwblaikie: Yes, Clang Static Analyzer checks may be run in Clang-tidy, if enabled by user.

**Comment by Karina5005** [comment]
I run it like this:
`clang-tidy test.cpp -checks=*, clang-analyzer-*`
So I expect that Clang Static Analyzer checks will be used too.

**Comment by haoNoQ** [comment]
The warning is emitted but hits a [common false positive suppression heuristic](https://github.com/llvm/llvm-project/blob/llvmorg-15-init/clang/lib/StaticAnalyzer/Core/BugReporterVisitors.cpp#L1194) that was designed for a completely different purpose. These heuristics badly need to be revisited but it's also a lot of work to properly evaluate them once they're lifted. Until then, it sounds like we're not great at catching null dereference bugs when the null pointer travels through a nested stack frame (such as, in this case, `get()`).
